prometheus/registry_test.go

```diff
@@ -746,38 +746,153 @@ func BenchmarkHandler(b *testing.B) {
 }
 
 func TestAlreadyRegistered(t *testing.T) {
-	reg := prometheus.NewRegistry()
 	original := prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Name:        "test",
+			Help:        "help",
+			ConstLabels: prometheus.Labels{"const": "label"},
+		},
+		[]string{"foo", "bar"},
+	)
+	equalButNotSame := prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Name:        "test",
+			Help:        "help",
+			ConstLabels: prometheus.Labels{"const": "label"},
+		},
+		[]string{"foo", "bar"},
+	)
+	originalWithoutConstLabel := prometheus.NewCounterVec(
 		prometheus.CounterOpts{
 			Name: "test",
 			Help: "help",
 		},
 		[]string{"foo", "bar"},
 	)
-	equalButNotSame := prometheus.NewCounterVec(
+	equalButNotSameWithoutConstLabel := prometheus.NewCounterVec(
 		prometheus.CounterOpts{
 			Name: "test",
 			Help: "help",
 		},
 		[]string{"foo", "bar"},
 	)
 	var err error
-	if err = reg.Register(original); err != nil {
-		t.Fatal(err)
-	}
-	if err = reg.Register(equalButNotSame); err == nil {
-		t.Fatal("expected error when registering equal collector")
-	}
-	if are, ok := err.(prometheus.AlreadyRegisteredError); ok {
-		if are.ExistingCollector != original {
-			t.Error("expected original collector but got something else")
+
+	t.Run("RegisterNormallyReregisterNormally", func(t *testing.T) {
+		reg := prometheus.NewRegistry()
+		if err = reg.Register(original); err != nil {
+			t.Fatal(err)
 		}
-		if are.ExistingCollector == equalButNotSame {
-			t.Error("expected original callector but got new one")
+
+		if err = reg.Register(equalButNotSame); err == nil {
+			t.Fatal("expected error when registering equal collector")
 		}
-	} else {
-		t.Error("unexpected error:", err)
-	}
+		if are, ok := err.(prometheus.AlreadyRegisteredError); ok {
```

Comment:
This if/else block is almost the same for all test maybe add a function to check it, something like `func checkAlreadyRegisteredError(t *testing.T, err error, original, equalButNotSame Collector) {...}`

Reply:
The whole test could (or should) be transformed into a table based one, filling in the registries used for registering, re-registering, and the CounterVecs to register and reregister.

In general, I don't mind code duplication in tests if it makes things very explicit. But perhaps at this point, the table approach is even easier to grok.

I'll work on it...

Reply:
The test is now table-driven.